### DISCUSSION TITLE
Missing something with Site-to-Site Case in Quickstart

### DISCUSSION THREAD

**Discussion by caffeineaddiction** [strongswan/strongswan#1916] [Q&A] [answered]
I have been tasked with setting up a "moon" in AWS to connect to another companies SonicWall Firewall "sun"

Through much trial and error I finally got it to connect ... but cant figure out how to make use of this newly established connection

So, strait out of the quickstart I have:
```
10.196.6.254/32 -- | 1.1.1.226 | === | 2.2.2.39 | -- 170.137.124.0/25
  moon-net              moon              sun          sun-net
```
  
```
$ sudo ipsec statusall
Status of IKE charon daemon (strongSwan 5.9.5, Linux 6.2.0-1011-aws, x86_64):
  uptime: 72 minutes, since Sep 19 19:32:35 2023
  malloc: sbrk 2134016, mmap 0, used 1417360, free 716656
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 9
  loaded plugins: charon aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm drbg attr kernel-netlink resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic counters
Listening IP addresses:
  172.31.43.251
  172.27.1.1
Connections:
       guava:  %any...2.2.2.39  IKEv1, dpddelay=30s
       guava:   local:  [1.1.1.226] uses pre-shared key authentication
       guava:   remote: [2.2.2.39] uses pre-shared key authentication
       guava:   child:  10.196.6.254/32 === 170.137.124.0/25 TUNNEL, dpdaction=restart
Security Associations (1 up, 0 connecting):
       guava[1]: ESTABLISHED 72 minutes ago, 172.31.43.251[1.1.1.226]...2.2.2.39[2.2.2.39]
       guava[1]: IKEv1 SPIs: f392a5d8ea132f70_i* 25b830a6dfafbd55_r, pre-shared key reauthentication in 22 hours
       guava[1]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
       guava{4}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: c0a9e056_i 59d03053_o
       guava{4}:  AES_CBC_256/HMAC_SHA2_256_128/MODP_2048, 0 bytes_i, 0 bytes_o, rekeying in 37 minutes
       guava{4}:   10.196.6.254/32 === 170.137.124.0/25

```

Best I can figure this means that the tunnel is established.  The other company (sun) sees me as connected job done.

The problem is even though it says that the tunnel is established, I can not ping `170.137.124.27` or telnet `170.137.124.27 11324` both of which the other company (sun) says I should be able to do.  Now when I have configured this with other VPNs like Tinc or OpenVPN ... I am use to there being a Tun0 and I have to setup routing tables accordingly.  However I see no virtual interface beyond the one I setup previously for tinc.

```
$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc fq_codel state UP group default qlen 1000
    link/ether 0e:4d:c0:87:42:cf brd ff:ff:ff:ff:ff:ff
    inet 172.31.43.251/20 metric 100 brd 172.31.47.255 scope global dynamic eth0
       valid_lft 2719sec preferred_lft 2719sec
    inet6 fe80::c4d:c0ff:fe87:42cf/64 scope link
       valid_lft forever preferred_lft forever
3: tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 500
    link/none
    inet 172.27.1.1/24 scope global tun0
       valid_lft forever preferred_lft forever
    inet6 fe80::c43d:614d:df1a:2046/64 scope link stable-privacy
       valid_lft forever preferred_lft forever
```

Both `/etc/strongswan.d/charon.conf` and the documentation seem to imply that a virtual ip will be created and routes assigned to said virtual ip ... but I seem to be missing something.

I attempted to setup an eth0:1 alias via ifconfig as well as `/etc/ipsec.d/vti-updown` in one of the guides ... and while it gives me an interface and I can setup a route for said interface ... the newly created interface seems to not be able to communicate with `170.137.124.27`

Here is  my current `/etc/ipsec.conf` w/ external IP addresses changed:
```
$ cat /etc/ipsec.conf
# /etc/ipsec.conf - strongSwan IPsec configuration file

config setup
    #charondebug="ike 2, knl 2, cfg 2, net 2, esp 2, dmn 2,  mgr 2"
    charondebug="ike 1, knl 1, cfg 1, net 1, esp 1, dmn 1, mgr 1"

conn %default
    ikelifetime=86400s
    keylife=3600s
    rekeymargin=3m
    keyingtries=1
    authby=secret
    keyexchange=ikev1

conn guava
    left=%any
    leftid=1.1.1.226
    leftsourceip=10.196.6.254/32
    leftsubnet=10.196.6.254/32

    right=2.2.2.39
    rightsubnet=170.137.124.0/25
    ike=aes256-sha256-modp2048!
    esp=aes256-sha256-modp2048!
    ikelifetime=86400s
    lifetime=3600s
    dpddelay=30s
    dpdtimeout=120s
    dpdaction=restart
    auto=start
```

It should be noted that Sun is fully expecting us to use a NAT on `10.196.6.254/32` and this connection is for us to reach from moon and other connections bridged with moon to sun.

Currently `sudo iptables-save` has nothing in it as firewall stuff has been configured through AWS ... and is pretty boring.  Basically just whitelist of my company and sun company.

What if anything am I doing wrong?  Should there be a Virtual IP?  Is the Virtual IP an IKEv2 thing?

**Comment by tobiasbrunner** [comment]
> I am use to there being a Tun0 and I have to setup routing tables accordingly. However I see no virtual interface beyond the one I setup previously for tinc.

That's not how policy-based IPsec works. You might want to read [the introduction](https://docs.strongswan.org/docs/5.9/howtos/introduction.html).

>     leftsourceip=10.196.6.254/32
>     leftsubnet=10.196.6.254/32

That's not how [virtual IPs](https://docs.strongswan.org/docs/5.9/features/vip.html) are configured. Either you configure `leftsourceip=%config` to request a virtual IP from the server (and you omit `leftsubnet` as that is automatically set to the assigned virtual IP), or, if that IP is statically assigned and not negotiated, you omit `leftsourceip` and leave `leftsubnet` as is but you also have to manually install that IP address on one of your local interfaces (could be on `lo`) before you start the connection.

> It should be noted that Sun is fully expecting us to use a NAT on `10.196.6.254/32` and this connection is for us to reach from moon and other connections bridged with moon to sun.

If you need a NAT (e.g. for hosts behind your client that are routed through it) then you need to install one yourself. Depending on how the virtual IP is assigned you can do that statically or dynamically via updown script.

**Comment by caffeineaddiction** [comment]
> That's not how policy-based IPsec works. You might want to read [the introduction](https://docs.strongswan.org/docs/5.9/howtos/introduction.html).

I read through that prior to posting, that and the page about virtual IPs but I guess I wasn't understanding it.

> IP is statically assigned and not negotiated, you omit `leftsourceip` and leave `leftsubnet` as is but you also have to manually install that IP address on one of your local interfaces (could be on lo) before you start the connection.

When you say "before"  you start the connection ... would the `leftupdown` script count as before or does it need to be manually configured prior to that even?

> If you need a NAT (e.g. for hosts behind your client that are routed through it) then you need to install one yourself. Depending on how the virtual IP is assigned you can do that statically or dynamically via updown script.

Is this a "This should/could be done in IPTables but is out of scope for this forum" type of thing?  Or is this a "This is a plugin or different program entirely but is out of s cope of this forum" type of thing?

**Reply by tobiasbrunner** [reply, marked as answer]
> would the `leftupdown` script count as before or does it need to be manually configured prior to that even?

The updown script is invoked after a CHILD_SA is negotiated and installed so that might be too late, unless you also want to install routes to the remote that reference the virtual IP as preferred source manually (if the IP is already there beforehand, strongSwan will install such routes for you in table 220). However, if you need a NAT this might not really matter as such a route wouldn't be necessary because the NAT will force the correct source IP to match the IPsec policies anyway. So you could theoretically install the IP and the NAT rule in an updown script if you wanted to.

To install the IP earlier, you could add it to an interface when the system starts via the system's network config (depends on the system/distro), or e.g. use the `charon.start-scripts` feature in strongswan.conf to do it when the daemon starts.


> Is this a "This should/could be done in IPTables but is out of scope for this forum" type of thing? Or is this a "This is a plugin or different program entirely but is out of s cope of this forum" type of thing?

Sure iptables is an option (or nft). And again, could be done in an updown or vici script (if the virtual IP is dynamic or you only want to NAT if the SA is up) or beforehand (if the IP is static and you don't want to mess with dynamic scripts).

**Reply by caffeineaddiction** [reply]
ty
